public/app/core/utils/dag.ts

```diff
@@ -189,6 +189,33 @@ export class Graph {
     return edges;
   }
 
+  descendants(nodes: Node[] | string[]): Node[] {
+    if (!nodes.length) {
+      return [];
+    }
+
+    if (isStringArray(nodes)) {
+      nodes = nodes.map((n) => this.nodes[n]).filter((n) => n !== undefined);
+    }
+
+    return this.descendantsRecursive(nodes);
+  }
+
+  private descendantsRecursive(nodes: Node[], descendants = new Set<Node>()): Node[] {
+    for (const node of nodes) {
+      const newDescendants = node.inputEdges
+        .map((n) => n.inputNode!)
+        .filter((n) => n !== undefined && !descendants.has(n));
+      for (const n of newDescendants) {
+        descendants.add(n);
```

Comment:
I think it's a little bit weird how this function mutates the `descendants` argument that's passed to it, and then spreads it into a new array to return it.

Maybe instead it could do `const newDescendants = new Set(descendants)` create a new Set at the beginning of function, add to that over the course of it, and then return that `newDescendants `?

Reply:
I think it would be fine for these functions to return Sets, because it looks like you convert it to a set anyway when you use it.